include/revng/PTML/ContextualCommentMap.h

```diff
@@ -0,0 +1,223 @@
+
+private:
+  // This assumes that `Trait::StatementType` is cheap (and, more importantly,
+  // safe) to copy
+  std::unordered_map<typename Trait::StatementType, CommentList> ResultMap;
+  CommentList HomelessComments;
+
+public:
+  ContextualCommentMap() = default;
+  ContextualCommentMap(const model::Function &Function, NodeType EntryNode) {
+    using LocationType = std::decay_t<typename Trait::LocationType>;
+    std::vector<std::pair<NumericScore, LocationType>> Scores;
+    Scores.resize(Function.Comments().size());
+    if (Scores.empty()) {
+      // No comments in this function, nothing to map.
+      return;
+    }
+
+    // First build a score map - select the best location for each comment
+    for (NodeType Node : llvm::depth_first(EntryNode)) {
+      for (const auto &Statement : Trait::getStatements(Node)) {
+        LocationType Location = Trait::getAddresses(Statement);
+
+        for (auto [Index, Comment] : llvm::enumerate(Function.Comments())) {
+
+          // Use the Tversky Index as the scoring function.
+          // Note the parameters controlling the weights of the sets: the higher
+          // a value is, the bigger the relative negative impact of
+          // the corresponding set having _extra_ elements compared to the other
+          // one.
+          //
+          // We want this because a line having unaccounted for addresses is a
+          // way bigger red flag that a (potentially user edited) comment having
+          // them.
+          auto Score = tverskyIndex<1, 2>(Comment.Location(), Location);
+
+          if (Score.has_value() && *Score > Scores[Index].first)
+            Scores[Index] = { *Score, Location };
+        }
+      }
+    }
+
+    // Then translate it into the mode-to-comment-list map taking into account
+    // the possibility of multiple nodes having the same location.
+    for (auto [Index, Comment] : llvm::enumerate(Function.Comments())) {
+      if (Scores[Index].first.Numerator == 0) {
+        // No feasible location for this comment, mark it as homeless.
+        HomelessComments.emplace_back(Index, false, &Comment.Location());
+
+      } else {
+        bool CommentWasPlaced = false;
+        for (NodeType Node : llvm::depth_first(EntryNode)) {
+          using StatementType = Trait::StatementType;
+          for (const StatementType &Statement : Trait::getStatements(Node)) {
+            if (Trait::getAddresses(Statement) == Scores[Index].second) {
+              bool IsLocationExact = std::ranges::equal(Comment.Location(),
+                                                        Scores[Index].second);
+              ResultMap[Statement].emplace_back(Index,
+                                                IsLocationExact,
+                                                &Comment.Location());
+              CommentWasPlaced = true;
+              break;
+            }
+          }
+
+          if (CommentWasPlaced)
+            break;
```

Comment:
Wait, `B` _and_ `D`? That could be very awkward in some cases with ties.

Ah, but wait, if we lock it to specific lists, it might actually be what we want! I'll need to think about this a bit more, but now I get why you were so adamant about domination earlier.

Reply:
Yeah, I'm aware of that, but as far as I remember, you mentioned that the plan is to duplicate the blocks _after_ the comments are already there.

Reply:
> we'll just run comment injection _after_ duplication

Apparently I cannot read :D My bad.